std/algorithm.d

```diff
@@ -1391,6 +1394,145 @@ unittest //13304
     static assert(is(typeof(reduce!((a, b)=>a+b)(data))));
 }
 
+
+// each
+/**
+Eagerly iterates over $(D r) and calls $(D fun) over _each element.
+
+Example:
+---
+void deleteOldBackups()
+{
+    import std.algorithm, std.datetime, std.file;
+    auto cutoff = Clock.currTime() - 7.days;
+    dirEntries("", "*~", SpanMode.depth)
+        .filter!(de => de.timeLastModified < cutoff)
+        .each!remove();
+}
+---
+
+If the range supports it, the value can be mutated in place. Examples:
+---
+arr.each!((ref a) => a++);
+arr.each!"a++";
+---
+
+If no predicate is specified, $(D each) will default to doing nothing
+but consuming the entire range. $(D .front) will be evaluated, but this
+can be avoided by explicitly specifying a predicate lambda with a
+$(D lazy) parameter.
+
+$(D each) also supports $(D opApply)-based iterators, so it will work
+with e.g. $(XREF parallelism, parallel).
+
+See_Also: $(XREF range,tee)
+
+ */
+template each(alias fun = "a")
+{
+    alias BinaryArgs = TypeTuple!(fun, "i", "a");
+
+    enum isRangeUnaryIterable(R) =
+        is(typeof(unaryFun!fun(R.init.front)));
+
+    enum isRangeBinaryIterable(R) =
+        is(typeof(binaryFun!BinaryArgs(0, R.init.front)));
+
+    enum isRangeIterable(R) =
+        isInputRange!R &&
+        (isRangeUnaryIterable!R || isRangeBinaryIterable!R);
+
+    enum isForeachUnaryIterable(R) =
+        is(typeof((R r) {
+            foreach (ref a; r)
+                cast(void)unaryFun!fun(a);
+        }));
+
+    enum isForeachBinaryIterable(R) =
+        is(typeof((R r) {
+            foreach (i, ref a; r)
+                cast(void)binaryFun!BinaryArgs(i, a);
+        }));
+
+    enum isForeachIterable(R) =
+        (!isForwardRange!R || isDynamicArray!R) &&
+        (isForeachUnaryIterable!R || isForeachBinaryIterable!R);
+
+    void each(Range)(Range range)
+    if (isRangeIterable!Range && !isForeachIterable!Range)
+    {
+        debug(each) pragma(msg, "Using while for ", Range.stringof);
+        static if (isRangeUnaryIterable!Range)
+        {
+            while (!range.empty)
+            {
+                cast(void)unaryFun!fun(range.front);
+                range.popFront();
+            }
+        }
+        else // if (isRangeBinaryIterable!Range)
+        {
+            size_t i = 0;
+            while (!range.empty)
+            {
+                cast(void)binaryFun!BinaryArgs(i, range.front);
+                range.popFront();
+                i++;
+            }
+        }
+    }
+
+    void each(Iterator)(Iterator iterator)
```

Comment:
I wasn't thinking of C++ iterators when I chose the term, but more of those found in Python/Ruby/C#. What term would you suggest, then? (If not for the parameter name, then for the inferred type name).